In [1]:
import numpy as np
import pandas as pd

In [2]:
df1 = pd.read_csv('w_data.csv', sep =";")
df1

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target  
0                40.0       49084    <=50K  
1                13.0       68553    <=50K  
2                40.0      143654    <=50K  
3                40.0       41510    <=50K  
4                40.0      195025    <=50K  
...               ...         ...      ...  
48838            36.0   735617153   <=50K.  
48839            40.0  1131807446   <=50K.  
48840            50.0   781422087   <=50K.  
48841            40.0   161771296   <=50K.  
48842            60.0  1344995623    >50K.  

[48843 rows x 7 columns]

In [3]:
df2 = pd.read_csv('p_data.csv', sep =";")
df2

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  
0       Bachelors           13.0        Never-married    Not-in-family  
1       Bachelors           13.0   Married-civ-spouse          Husband  
2         HS-grad            9.0             Divorced    Not-in-family  
3            11th            7.0   Married-civ-spouse          Husband  
4       Bachelors           13.0   Married-civ-spouse             Wife  
...           ...            ...                  ...              ...  
48837   Bachelors           13.0        Never-married        Own-child  
48838   Bachelors           13.0             Divorced    Not-in-family  
48839     HS-grad            9.0              Widowed   Other-relative  
48840   Bachelors           13.0   Married-civ-spouse          Husband  
48841   Bachelors           13.0             Divorced        Own-child  

[48842 rows x 9 columns]

1- Cuantos ID tienen en común. Nombrar cuales no tienen en común

In [4]:
df = pd.merge(df1, df2, on="ID")
df

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48837            Private  245211.0           0.0           0.0   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   

       Hours per week          ID   Target   Age      Sex  \
0                40.0       49084    <=50K  39.0     Male   
1                13.0       68553    <=50K  50.0     Male   
2                40.0      143654    <=50K  38.0     Male   
3                40.0       41510    <=50K  53.0     Male   
4                40.0      195025    <=50K  28.0   Female   
...               ...         ...      ...   ...      ...   
48837            40.0   597186815   <=50K.  33.0     Male   
48838            36.0   735617153   <=50K.  39.0   Female   
48839            40.0  1131807446   <=50K.  64.0     Male   
48840            50.0   781422087   <=50K.  38.0     Male   
48841            40.0   161771296   <=50K.  44.0     Male   

                      Race         Country   Education  Education-Num  \
0                    White   United-States   Bachelors           13.0   
1                    White   United-States   Bachelors           13.0   
2                    White   United-States     HS-grad            9.0   
3                    Black   United-States        11th            7.0   
4                    Black            Cuba   Bachelors           13.0   
...                    ...             ...         ...            ...   
48837                White   United-States   Bachelors           13.0   
48838                White   United-States   Bachelors           13.0   
48839                Black   United-States     HS-grad            9.0   
48840                White   United-States   Bachelors           13.0   
48841   Asian-Pac-Islander   United-States   Bachelors           13.0   

            Marital Status     Relationship  
0            Never-married    Not-in-family  
1       Married-civ-spouse          Husband  
2                 Divorced    Not-in-family  
3       Married-civ-spouse          Husband  
4       Married-civ-spouse             Wife  
...                    ...              ...  
48837        Never-married        Own-child  
48838             Divorced    Not-in-family  
48839              Widowed   Other-relative  
48840   Married-civ-spouse          Husband  
48841             Divorced        Own-child  

[48842 rows x 15 columns]

In [5]:
len(df1.value_counts('ID'))

48843

In [6]:
len(df2.value_counts('ID'))

48842

In [7]:
len(df.value_counts('ID'))

48842

In [8]:
df0 = pd.merge(df1,df2, how='left', on="ID", indicator=True)
df0

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target   Age      Sex  \
0                40.0       49084    <=50K  39.0     Male   
1                13.0       68553    <=50K  50.0     Male   
2                40.0      143654    <=50K  38.0     Male   
3                40.0       41510    <=50K  53.0     Male   
4                40.0      195025    <=50K  28.0   Female   
...               ...         ...      ...   ...      ...   
48838            36.0   735617153   <=50K.  39.0   Female   
48839            40.0  1131807446   <=50K.  64.0     Male   
48840            50.0   781422087   <=50K.  38.0     Male   
48841            40.0   161771296   <=50K.  44.0     Male   
48842            60.0  1344995623    >50K.   NaN      NaN   

                      Race         Country   Education  Education-Num  \
0                    White   United-States   Bachelors           13.0   
1                    White   United-States   Bachelors           13.0   
2                    White   United-States     HS-grad            9.0   
3                    Black   United-States        11th            7.0   
4                    Black            Cuba   Bachelors           13.0   
...                    ...             ...         ...            ...   
48838                White   United-States   Bachelors           13.0   
48839                Black   United-States     HS-grad            9.0   
48840                White   United-States   Bachelors           13.0   
48841   Asian-Pac-Islander   United-States   Bachelors           13.0   
48842                  NaN             NaN         NaN            NaN   

            Marital Status     Relationship     _merge  
0            Never-married    Not-in-family       both  
1       Married-civ-spouse          Husband       both  
2                 Divorced    Not-in-family       both  
3       Married-civ-spouse          Husband       both  
4       Married-civ-spouse             Wife       both  
...                    ...              ...        ...  
48838             Divorced    Not-in-family       both  
48839              Widowed   Other-relative       both  
48840   Married-civ-spouse          Husband       both  
48841             Divorced        Own-child       both  
48842                  NaN              NaN  left_only  

[48843 rows x 16 columns]

Tienen 48842 ID en común, uno solo no tienen en común: el ID= 1344995623

In [9]:
df_exclude = df0[df0['_merge'] == 'left_only']
df_exclude

Workclass    fnlwgt  Capital Gain  Capital Loss  Hours per week  \
48842   Self-emp-inc  182148.0           0.0           0.0            60.0   

               ID  Target  Age  Sex Race Country Education  Education-Num  \
48842  1344995623   >50K.  NaN  NaN  NaN     NaN       NaN            NaN   

      Marital Status Relationship     _merge  
48842            NaN          NaN  left_only

2- Corregir los target a solo dos valores distintos ‘<=50k’ y ‘>50k’

In [10]:
df["Target"].value_counts()

 <=50K     24720
 <=50K.    12435
 >50K       7841
 >50K.      3845
Name: Target, dtype: int64

In [11]:
df0 = df0.replace(to_replace="<=50K", value="<=50k", regex=True)
df0 = df0.replace(to_replace="<=50k.", value="<=50k", regex=True)
df0 = df0.replace(to_replace=">50K", value=">50k", regex=True)
df0= df0.replace(to_replace=">50k.", value=">50k", regex=True)

In [31]:
valores = df0["Target"].value_counts()
valores

 <=50k    37155
 >50k     11687
Name: Target, dtype: int64

3- Calcular la proporción que hay entre personas con salarios >50k respecto del total para Entre personas de distinto sexo, Raza y Educacion. Por ejemplo cuál es el porcentaje de mujeres con un salario mayor a 50K y compararlo con el de los hombres. Lo mismo para personas de distintas razas y Educación ( group by para cada una de las condiones)

In [30]:
genero= df0["Target"].groupby([df0["Sex"], df0["Target"]]).count()
genero

Sex      Target
 Female   <=50k    14423
          >50k      1769
 Male     <=50k    22732
          >50k      9917
Name: Target, dtype: int64

In [19]:
genero.unstack(level="Sex") 

Sex      Female   Male
Target                
 <=50k    14423  22732
 >50k      1769   9917

In [29]:
educacion = df0["Target"].groupby([df0["Education"], df0["Target"]]).count()
educacion

Education      Target
 10th           <=50k     1302
                >50k        87
 11th           <=50k     1720
                >50k        92
 12th           <=50k      609
                >50k        48
 1st-4th        <=50k      239
                >50k         8
 5th-6th        <=50k      482
                >50k        27
 7th-8th        <=50k      893
                >50k        62
 9th            <=50k      715
                >50k        41
 Assoc-acdm     <=50k     1188
                >50k       413
 Assoc-voc      <=50k     1539
                >50k       522
 Bachelors      <=50k     4712
                >50k      3312
 Doctorate      <=50k      163
                >50k       431
 HS-grad        <=50k    13281
                >50k      2503
 Masters        <=50k     1198
                >50k      1459
 Preschool      <=50k       82
                >50k         1
 Prof-school    <=50k      217
                >50k       617
 Some-college   <=50k     8815
                >

In [28]:
race = df0["Target"].groupby([df0["Race"], df0["Target"]]).count()
race

Race                 Target
 Amer-Indian-Eskimo   <=50k      415
                      >50k        55
 Asian-Pac-Islander   <=50k     1110
                      >50k       409
 Black                <=50k     4119
                      >50k       566
 Other                <=50k      356
                      >50k        50
 White                <=50k    31155
                      >50k     10606
Name: Target, dtype: int64